File: library/src/main/java/net/robinx/queue/JobHandler.java
package net.robinx.queue;

/**
 * Created by Robin on 2016/6/28 17:40.
 */
public class JobHandler<T extends Job<T>> {

    private JobQueue<T> mJobQueue;

    private int mThreadPoolSize;

    private Interceptor<T> mInterceptor;

    private JobHandler(Builder<T> builder){
        this.mThreadPoolSize = builder.threadPoolSize;
        this.mInterceptor = builder.interceptor;

        if (mThreadPoolSize == 0) {
            mThreadPoolSize = JobQueue.DEFAULT_THREAD_POOL_SIZE;
        }
        if (mInterceptor == null) {
            mInterceptor = new DefaultInterceptor<>();
        }

        mJobQueue = new JobQueue(mThreadPoolSize,mInterceptor);
        mJobQueue.start();
    }

    public JobQueue<T> enqueue(T element){

        return mJobQueue.add(element);
    }

    public <O> JobQueue<T> cancelAll(O tag){
        mJobQueue.cancelAll(tag);
        return mJobQueue;
    }

    public static class Builder<T extends Job<T>>{
        private int threadPoolSize;
        private Interceptor<T> interceptor;

        public Builder threadPoolSize(int threadPoolSize) {
            this.threadPoolSize = threadPoolSize;
            return this;
        }

        public Builder interceptor(Interceptor<T> interceptor) {
            this.interceptor = interceptor;
            return this;
        }

        public JobHandler<T> build(){
            return new JobHandler<>(this);
        }
    }
}

File: library/src/main/java/com/github/jobqueue/JobHandler.kt
package com.github.jobqueue

/**
 * Created by Robin on 2016/6/28 17:40.
 */
class JobHandler<T : Job<T>> private constructor(builder: Builder<T>) {
    private val mJobQueue: JobQueue<T>

    private var mThreadPoolSize: Int

    private var mInterceptor: Interceptor<T>?

    init {
        this.mThreadPoolSize = builder.threadPoolSize
        this.mInterceptor = builder.interceptor

        if (mThreadPoolSize == 0) {
            mThreadPoolSize = JobQueue.DEFAULT_THREAD_POOL_SIZE
        }
        if (mInterceptor == null) {
            mInterceptor = DefaultInterceptor()
        }

        mJobQueue = JobQueue<T>(mThreadPoolSize, mInterceptor)
        mJobQueue.start()
    }

    fun enqueue(element: T): JobQueue<T> {
        return mJobQueue.add(element)
    }

    fun <O> cancelAll(tag: O): JobQueue<T> {
        mJobQueue.cancelAll(tag)
        return mJobQueue
    }

    class Builder<T : Job<T>> {
        internal var threadPoolSize: Int = 0
        var interceptor: Interceptor<T>? = null

        fun threadPoolSize(threadPoolSize: Int): Builder<T> {
            this.threadPoolSize = threadPoolSize
            return this
        }

        fun interceptor(interceptor: Interceptor<T>?): Builder<T> {
            this.interceptor = interceptor
            return this
        }

        fun build(): JobHandler<T> {
            return JobHandler(this)
        }
    }
}
